### PR TITLE
oauth support, add the token in a configuration specified cookie

### PR DESCRIPTION
**Pull request by bmccandless**
Previously, the id token was stored in the session token.
Now, it can be placed in a different cookie with different properties.

### CODE REVIEW

File: server/auth/auth_oauth.py [hunk 1]
```diff
@@ -121,10 +142,24 @@ def logout(self):
     def callback(self):
         token = self.client.authorize_access_token()
         id_token = token.get("id_token")
-        session[self.CXG_ID_TOKEN] = id_token
-        del session["oauth_callback_redirect"]
         oauth_callback_redirect = session.get("oauth_callback_redirect", "/")
+        del session["oauth_callback_redirect"]
```
Comment (maniarathi) [inline server/auth/auth_oauth.py:146]: Can be more precisely refactored to:
`oauth_callback_redirect = session.pop("oauth_callback_redirect", "/")`

File: server/auth/auth_oauth.py [hunk 2]
```diff
@@ -121,10 +142,24 @@ def logout(self):
...
+            args = self.cookie_params.copy()
+            del args["key"]
+            try:
+                resp.set_cookie(
+                    self.cookie_params["key"],
+                    id_token,
+                    **args)
+                g.token = id_token
+            except Exception as e:
+                raise AuthenticationError(f"unable to set_cookie {self.cookie_params}") from e
```
[the remark this answers is not in the document]
Comment (bmccandless) [reply server/auth/auth_oauth.py:161]: It's for exception chaining which, will provide better logging of the error.

File: server/auth/auth_oauth.py [hunk 3]
```diff
@@ -24,15 +24,29 @@ class AuthTypeOAuth(AuthTypeClientBase):
...
-        self.client_secret = app_config.authentication__params_oauth__client_secret
-        self.callback_base_url = app_config.authentication__params_oauth__callback_base_url
+        self.api_base_url = server_config.authentication__params_oauth__api_base_url
+        self.client_id = server_config.authentication__params_oauth__client_id
+        self.client_secret = server_config.authentication__params_oauth__client_secret
+        self.callback_base_url = server_config.authentication__params_oauth__callback_base_url
+        self.session_cookie = server_config.authentication__params_oauth__session_cookie
+        if not self.session_cookie:
+            self.cookie_params = server_config.authentication__params_oauth__cookie
+            if type(self.cookie_params) != dict:
```
[the remark this answers is not in the document]
Comment (bmccandless) [reply server/auth/auth_oauth.py]: fixed

File: server/auth/auth_oauth.py [hunk 4]
```diff
@@ -24,15 +24,29 @@ class AuthTypeOAuth(AuthTypeClientBase):
...
+                raise ConfigurationError("either session_cookie or cookie must be set")
+            valid_keys = set(("key", "max_age", "expires", "path", "domain", "secure", "httponly", "samesite"))
```
[the remark this answers is not in the document]
Comment (bmccandless) [reply server/auth/auth_oauth.py]: fixed

File: server/auth/auth_oauth.py [hunk 5]
```diff
@@ -24,15 +24,29 @@ class AuthTypeOAuth(AuthTypeClientBase):
...
+            keys = set(self.cookie_params.keys())
+            unknown = keys - valid_keys
+            if unknown:
+                raise ConfigurationError(f"unexpected key in cookie params: {', '.join(unknown)}")
+            if "key" not in keys:
+                raise ConfigurationError("must have a key (name) in the cookie params")
```
[the remark this answers is not in the document]
Comment (bmccandless) [reply server/auth/auth_oauth.py]: fixed

File: server/auth/auth_oauth.py (hunk 1 above)
Comment (bmccandless) [reply server/auth/auth_oauth.py]: fixed

**Pull request merged**